File: src/bin/0347-top-k-frequent-elements.cpp
// @leetup=info id=347 lang=cpp slug=top-k-frequent-elements

#include <algorithm>
#include <cassert>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
    std::map<int, size_t> count;
    for (int n : nums) {
      ++count[n];
    }
    std::priority_queue<std::tuple<size_t, int>> q;
    for (auto [key, cnt] : count) {
      q.push({-cnt, key});
      if (q.size() > static_cast<size_t>(k)) {
        q.pop();
      }
    }
    std::vector<int> res;
    while (!q.empty()) {
      auto [_cnt, k] = q.top();
      q.pop();
      res.push_back(k);
    }
    return res;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0347-top-k-frequent-elements.rs
// @leetup=info id=347 lang=rust slug=top-k-frequent-elements

// @leetup=code
use std::collections::*;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut count = HashMap::new();
        for n in nums {
            *count.entry(n).or_insert(0) += 1;
        }
        let mut q = BinaryHeap::new();
        for (n, count) in count {
            q.push((-count, n));
            if q.len() > k as usize {
                q.pop();
            }
        }
        q.into_iter().map(|(_, n)| n).collect()
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::top_k_frequent;
}
